qmail-autoresponder.1
.\" $Id$
.TH qmail-autoresponder 1
.SH NAME
qmail-autoresponder \- Produce automatic responses with qmail
.SH SYNOPSIS
.B qmail-autoresponder
[
.B \-cLNqT
] [
.BR "-O NAME" [ =VALUE ]
] [
.B -n NUM
] [
.B -s STR
] [
.B -t TIME
] [
.I MESSAGE-FILE
]
.I DIRECTORY
.SH DESCRIPTION
.B qmail-autoresponder
sends an automatically-generated message in response to an incoming
message.
It also tracks the number of responses it sends to each sender in
order to limit the rate at which it sends responses, to prevent mail
flooding or denial of service attacks.
By default, it limits the number of responses it sends to a
maximum of one message per hour per sender.  It will not respond to
any known mailing list, nor will it respond to either bounces or
loops.
.P
The
.I MESSAGE-FILE
is used as the response message, and must contain a complete mail
message including
.I From:
and
.I Subject:
headers.
If the
.I MESSAGE-FILE
argument is not present, it defaults to
.IR message.txt .
All occurrences of
.I %S
in the message file are replaced with the original message's subject,
.I %s
with the original message's sender (to which the response will be sent), and
.I %r
with the original message's recipient.
To put a percent sign in the message, write it as
.IR %% .
.SS RESPONSE COUNTING
For each response
.B qmail-autoresponder
sends, it creates a file in
.IR DIRECTORY .
The name of that file consists of the current UNIX time number, a period,
the process ID of
.BR qmail-autoresponder ,
a period, and the envelope sender address (with any
.I /
characters replaced with
.I :
to prevent creation of files outside of
.IR DIRECTORY ).
When it receives a message, it scans
.IR DIRECTORY .
Any files that are older than the time interval (see below) are
deleted and ignored.  If the number of remaining files with the same
sender address is greater than or equal to the maximum number of
replies, no response is generated.
.SS LOGGING
If
.I DIRECTORY
contains a writeable file named
.IR log.txt ,
then a line is added to this file each time a sender is processed.  The
format of these lines are as follows:
.RS
.EX
TIMESTAMP +SENDER
TIMESTAMP -SENDER
.EE
.RE
The first line indicates a message was sent to the sender, and the
second that a message was not sent.  The timestamp is expressed as
standard UNIX time (the number of seconds since Jan 1 1970 GMT).
.SH OPTIONS
.B qmail-autoresponder
supports options specified on the command line, in a file named
.BR config.txt ,
and as individual files in
.IR DIRECTORY .
If
.B config.txt
is present, any individual config files are ignored.
Option files in the directory (either
.B config.txt
or individual files) override options set on the command line.
.P
The
.B config.txt
file contains one setting per line, with the option name separated from the value by a
.IR = .
The value extends from the first character after the
.I =
to the end of the line.
Lines starting with
.I #
are ignored.
.P
The message may also be placed in this configuration file. It is
separated from the settings by a single blank line. If present,
.B message.txt
is not read for the response message.
.P
For example:
.RS
.EX
copymsg=1
# Keep all the X-* headers along with the subject.
headerkeep=subject:x-*
bcc=sendmeacopy@example.com
starttime=2018-03-14 15:59:00
subject=This is a very long response subject that does not include the original

From: Automatic Responder <recipient@example.com>
Subject: Vacation response (%S)

This is an automated response to your email...
.EE
.RE
.SS "COMMAND LINE"
.TP
.B -c
Equivalent to
.B -O copymsg=1
.TP
.B -L
Create new files (instead of making links) for each recipient file.
This uses more inodes, but may be required for filesystems (such as
Coda) that don't handle links properly.
.TP
.B -N
Write the generated response to standard output instead of sending it.
.TP
.B -n NUM
Equivalent to
.B -O msglimit=NUM
.TP
.BR "-O NAME" [ =VALUE ]
Sets named option
.B NAME
to
.BR VALUE .
If
.B VALUE is omitted, a value of
.B 1
is used.  See below for all the named options.
.TP
.B -q
Quiet.  Don't show error messages.
.TP
.B -s STR
Equivalent to
.B -O subject_prefix=STR
.TP
.B -T
Do not add a
.I To: <SENDER>
line to the response.
.TP
.B -t TIME
Equivalent to
.B -O timelimit=TIME
.SS "NAMED OPTIONS"
.TP
.B bcc \fR(string)
If given, all responses are copied to this address in addition to the
original sender.
.TP
.B copymsg \fR(boolean)
Copy the original message into the response.
.TP
.B endtime \fR(string)
Set the ending time for which this autoresponder is active.
Messages received after this time will not cause a response to be generated.
.TP
.B headerkeep \fR(string)
If this option is given, only the headers listed in the value are copied
from the message into the response.  Separate header names with a colon.
If neither this nor
.B headerstrip
are used, all headers are copied.
.TP
.B headerstrip \fR(string)
If this option is given and
.B headerkeep
is not, the headers listed in the value are omitted when copying from
the message into the response.  Separate header names with a colon.
Header names are treated as patterns, and so may contain
.I *
to match any string.  For example, the pattern
.I x-*
will match the headers
.I X-DomainKeys:
and
.IR X-SpamLevel: .
.TP
.B msglimit \fR(integer)
Set the maximum number of replies per
.B timelimit
(defaults to 1).
.TP
.B no_inreplyto \fR(boolean)
Do not add an
.I In-Reply-To:
header to the outgoing message.  Normally, if a
.I Message-Id:
header is found in the original message,
.B qmail-autoresponder
will add the contents of that header to the outgoing message.  This
allows some email programs to group responses more intelligently.
.TP
.B numlines \fR(integer)
Limits the maximum number of lines copied from the original message body
into the response.  This option is only relevant if the
.B copymsg
option is also used.  If not given, all the lines in the original
message body are copied.  This limit does not impact how many lines of
the header are copied.
.TP
.B separator \fR(string)
Add a separator line (or lines) between the response message and the
copied message.
.TP
.B starttime \fR(string)
Set the starting time for which this autoresponder is active.
Messages received before this time will not cause a response to be generated.
.TP
.B subject_prefix \fR(string)
Add a subject line to the response message composed of the prefix
.I STR
followed by the original subject.  This can also be accomplished by
putting the following line at the start of the message file:

.EX
Subject: STR %S
.EE
.TP
.B timelimit \fR(integer)
Set the interval during which responses are counted (in seconds)
(defaults to 1 hour).
.SH RETURN VALUE
Exits zero if there were no problems, otherwise exits 111.
.SH ENVIRONMENT
.TP
.B DTLINE
.B qmail-autoresponder
searches for this line (generated by qmail-local) in the input and
bounces the message if it is present (which indicates a mail loop).
It then adds the line to the output.
.TP
.B SENDER
This contains the envelope sender address and is used as the address
to send the reply to.


NEWS
-------------------------------------------------------------------------------
Changes in version 2.0

- Changed the filenames of response record files so they sort by time in
  a directory listing. This may cause extra responses after upgrading.

- Added support for putting the original sender (%s) and recipient (%r)
  in the response message.

- Added support for single file config mode, including the response.

- Fixed handling of MySQL default timestamp value.
-------------------------------------------------------------------------------
Changes in version 1.0

- Added options to control the start and end time for autoresponders.

- Added errors for invalid configuration options.
-------------------------------------------------------------------------------
Changes in version 0.98

- Manually invoke qmail-inject and qmail-queue separately in order to
  fully control the envelope and capture the qmail PID.

- Added a "bcc" option to allow for copying autoresponses to an address
  other than the sender.

- Add an optional separator before the copied message.

- Add a newline after the response message and after the copied header,
  to make sure there is at least one blank line between them.

- The MySQL autoresponder now supports an optional table name prefix to
  better interoperate with other applications using the same database.

- Added support for a log file or table that is not pruned.

- Updated for bglibs v2.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.97

- Completely rewrote header handling, improving its handling of headers
  split across multiple lines among other things.

- Added support for omitting parts of the copied headers.

- Internal MIME headers are now stripped from multipart messages.

- Added support for limiting the number of lines copied from the
  original message body.

- Added support for setting autoresponder options through files.

- Added an option to control the generation of the In-Reply-To: header.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96.2

- Fixed up the multi-line strings that no longer compile with recent
  compilers.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96.1

- Added missing MySQL schema to package.

- Switched to new bglibs (1.006) library usage.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.96

- Added code to use MySQL instead of files for all the autoresponse
  data.

- Removed the option to omit the "To:" header, to allow for copying the
  response to extra addresses by adding BCCs to the message.

- Switched to using external bglibs.

- Removed adding the Delivered-To: line to outgoing messages.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.95

- Made the "MESSAGE-FILE" parameter optional, defaulting to
  "message.txt" in the autoresponder directory.

- Added a patch from Erik Sjoelund <erik@xpedio.com> to insert the
  original message's message-ID into a new In-Reply-To header field.

- Ignore messages containing the header fields listed in RFC 2369:
	List-(Help|Unsubscribe|Subscribe|Post|Owner|Archive):

- Portability fixes and documentation bugs.

- Added a manual page.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.93

- Fixed an off-by-one bug in the rate limiting logic (again).

- Fixed an off-by-one bug in the Delivered-To checking code.

- The output writing routine will now substitute "%S" in the
  autoresponse message with the original message's subject.
-------------------------------------------------------------------------------
Changes in version 0.92

- Fixed the rate limiting logic that would prevent messages from being
  sent out at the appropriate interval.

- Fixed bug in option parsing.
-------------------------------------------------------------------------------
Changes in version 0.91

- Added an option to insert the original subject into the reply.

- Simplified the header scanning logic.
-------------------------------------------------------------------------------


qmail-autoresponder.c
#include <bglibs/sysdeps.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <bglibs/fmt.h>
#include <bglibs/ibuf.h>
#include <bglibs/wrap.h>
#include <bglibs/str.h>
#include "qmail-autoresponder.h"

const char usage_args[] = "[MESSAGE-FILE] DIRECTORY";
const char usage_post[] =
"  Temporary files are put into DIRECTORY track senders' rates.\n"
"  MESSAGE-FILE defaults to 'message.txt'.\n";

static const char* opt_msgfilename = "message.txt";
static const char* opt_logfilename = "log.txt";
static const char* opt_configfilename = "config.txt";
static str safe_sender;
static const char* last_filename = 0;

static void read_message(const char* filename)
{
  int fd;
  long rd;
  char buf[4096];
  if ((fd = open(filename, O_RDONLY)) == -1)
    usage("Could not open message file.");
  while ((rd = read(fd, buf, sizeof buf)) > 0)
    if (!str_catb(&response, buf, rd))
      fail_temp("Out of memory while reading in message file.");
  close(fd);
}

static void read_config_files(void)
{
  struct option* option;
  int fd;
  char buf[4096];
  size_t rd;
  char* end;
  
  for (option = options; option->name != 0; ++option) {
    /* Open and read all option files */
    if ((fd = open(option->name, O_RDONLY)) == -1) {
      if (errno == ENOENT)
	continue;
      fail_temp("System error opening configuration files.");
    }
    if ((rd = read(fd, buf, sizeof buf - 1)) == (size_t)-1)
      fail_temp("System error reading configuration files.");
    close(fd);
    /* Make sure it's NUL terminated and strip off all but the first line. */
    buf[rd] = 0;
    if ((end = strchr(buf, '\n')) != 0) {
      *end = 0;
      rd = end - buf;
    }
    /* Parse the data */
    option->copyfn(option->ptr, buf, rd);
  }
}

static int read_config_file(void)
{
  ibuf in;
  str line = {0};

  if (!ibuf_open(&in, opt_configfilename, 0)) {
    if (errno == ENOENT)
      return 0;
    fail_temp("Could not read config.txt");
  }
  while (ibuf_getstr(&in, &line, '\n')) {
    /* Strip just the trailing \n */
    str_truncate(&line, line.len - 1);
    if (line.len == 0)
      break;
    if (line.s[0] == '#')
      continue;
    /* Match lines like "key=value" */
    unsigned eq = str_findfirst(&line, '=');
    if ((int)eq > 0 && eq < line.len) {
      line.s[eq++] = 0;
      handle_option(line.s, line.s + eq, line.len - eq);
    }
  }
  str_free(&line);

  if (!ibuf_readall(&in, &response))
    fail_temp("Error while reading message from config file.");

  ibuf_close(&in);
  return 1;
}

void init_autoresponder(int argc, char* argv[])
{
  int i = 0;
  switch (argc) {
  case 0: usage("Too few command-line arguments.");
  case 1: break;
  case 2: opt_msgfilename = argv[i++]; break;
  default: usage("Too many command-line arguments.");
  }
  if(chdir(argv[i]) == -1)
    usage("Could not change directory to DIRECTORY.");
  if (!read_config_file())
    read_config_files();
  if (response.len == 0)
    read_message(opt_msgfilename);
}

static void create_link(char* filename)
{
  int fd;
  
  /* Conserve inodes -- create links when possible */
  if(last_filename)
    if(link(last_filename, filename) == 0)
      return;

  /* Otherwise, create a new 0-byte file now */
  fd = open(filename, O_WRONLY|O_CREAT|O_EXCL, 0444);
  if(fd == -1)
    fail_temp("Unable to create temporary file for sender.");
  close(fd);
}

int count_history(const char* sender)
{
  DIR* dir = opendir(".");
  direntry* entry;
  unsigned count = 0;

  /* Translate all '/' to ':', to avoid fake paths in email addresses */
  wrap_str(str_copys(&safe_sender, sender));
  str_subst(&safe_sender, '/', ':');

  /* check if there are too many responses in the logs */
  while((entry = readdir(dir)) != NULL) {
    char* end;
    time_t message_time;
    unsigned message_pid;
    if(entry->d_name[0] == '.')
      continue;
    message_time = strtoul(entry->d_name, &end, 10);
    if(message_time == 0 || end == NULL || *end != '.')
      continue;
    if((unsigned long)(now - message_time) > opt_timelimit) {
      /* too old..ignore errors on unlink */
      if (!opt_nodelete)
	unlink(entry->d_name);
    } else {
      message_pid = strtoul(end+1, &end, 10);
      if(message_pid == 0 || end == NULL || *end != '.')
        continue;
      if(strcasecmp(end+1, safe_sender.s)==0)
	/* If the user's count is already over the max,
	 * don't record any more. */
	if(++count >= opt_msglimit)
	  return 0;
      last_filename = entry->d_name;
    }
  }

  return 1;
}

void log_sender(const char* sender, int responded)
{
  int fd;

  if (responded) {
    /* create the filename, format "TIME.PID.SENDER" */
    /* The PID is added to avoid collisions. */
    str_copyf(&tmpstr, "lu\\.u\\.s", now, getpid(), safe_sender.s);
    create_link(tmpstr.s);
  }

  if ((fd = open(opt_logfilename, O_WRONLY | O_APPEND)) != -1) {
    str_copyf(&tmpstr, "u{ }cs{\n}", now, (responded ? '+' : '-'), sender);
    if (write(fd, tmpstr.s, tmpstr.len) != tmpstr.len) { /* ignore */ }
    close(fd);
  }
  (void)sender;
}
